allennlp/modules/span_extractors/bidirectional_endpoint_span_extractor.py

```diff
@@ -0,0 +1,230 @@
+        forward_sequence = forward_sequence.contiguous()
+        backward_sequence = backward_sequence.contiguous()
+
+        # shape (batch_size, num_spans)
+        span_starts, span_ends = [index.squeeze(-1) for index in indices.split(1, dim=-1)]
+
+        if indices_mask is not None:
+            span_starts *= indices_mask
+            span_ends *= indices_mask
+
+        # We want `exclusive` span starts, so we remove 1 from the forward span starts
+        # as the AllenNLP ``SpanField`` is inclusive.
+        # shape (batch_size, num_spans)
+        exclusive_span_starts = span_starts - 1
+        # shape (batch_size, num_spans)
+        start_sentinel_mask = (exclusive_span_starts == -1).float().unsqueeze(-1)
+
+        # We want `exclusive` span ends for the backward direction
+        # (so that the `start` of the span in that direction is exlusive), so
+        # we add 1 to the span ends as the AllenNLP ``SpanField`` is inclusive.
+        exclusive_span_ends = span_ends + 1
+
+        if sequence_mask is not None:
+            # shape (batch_size)
+            sequence_lengths = get_lengths_from_binary_sequence_mask(sequence_mask)
+        else:
+            # shape (batch_size), filled with the sequence length size of the sequence_tensor.
+            sequence_lengths = ones_like(sequence_tensor[:, 0, 0]).long() * sequence_tensor.size(1)
+
+        # shape (batch_size, num_spans)
+        end_sentinel_mask = (exclusive_span_ends == sequence_lengths.unsqueeze(-1)).long()
+
+        # Forward Direction: start indices are exclusive. Shape (batch_size, num_spans, input_size / 2)
+        forward_start_embeddings = batched_index_select(forward_sequence, exclusive_span_starts)
+        # Forward Direction: end indices are inclusive, so we can just use span_ends.
+        # Shape (batch_size, num_spans, input_size / 2)
+        forward_end_embeddings = batched_index_select(forward_sequence, span_ends)
+
+        # Backward Direction: start indices are actually the end indices, because we
+        # are going backwards, which we want to be inclusive, so we can just use span_starts.
+        # Shape (batch_size, num_spans, input_size / 2)
+        backward_start_embeddings = batched_index_select(backward_sequence, span_starts)
+        # Backward Direction: end indices are actually the start indices, which we want to
+        # be exclusive. As we added 1 to the span_ends to make them exclusive, which might have
+        # caused indices equal to the sequence_length to become out of bounds, we multiply
+        # by the inverse of the end_sentinel mask to erase these indices (as we will replace
+        # them anyway in the block below).
+        # Shape (batch_size, num_spans, input_size / 2)
+        backward_end_embeddings = batched_index_select(backward_sequence,
+                                                       exclusive_span_ends * (1 - end_sentinel_mask))
+        if self._use_sentinels:
```

Comment:
Are you planning on crashing anywhere if you're not using sentinels, and you give an out-of-bounds span index?  We talked about that at one point, but I don't see it in here.  Looks like you'll just silently get really weird behavior.